File: backend/src/sql/documents.ts
export const DOCUMENT_FIELDS = `
  id, org_id, connector_id, external_id, url, title, kind, ext,
  author_name, author_email, content_hash, checksum, content_type, source_path,
  source_last_modified_at, source_permissions, extraction_status,
  extraction_error_code, extraction_version, chunking_version, indexing_version,
  word_count, mtime, indexed_at, metadata
`;

export function buildListDocumentsSql(filters: string[], limitParam: number, offsetParam: number): string {
  return `
    SELECT ${DOCUMENT_FIELDS}
    FROM documents
    WHERE ${filters.join(" AND ")}
    ORDER BY indexed_at DESC
    LIMIT $${limitParam} OFFSET $${offsetParam}
  `;
}

export function buildCountDocumentsSql(filters: string[]): string {
  return `
    SELECT count(*)::int AS total
    FROM documents
    WHERE ${filters.join(" AND ")}
  `;
}

export const SQL_SELECT_DOCUMENT_BY_ID = `
  SELECT ${DOCUMENT_FIELDS}
  FROM documents
  WHERE id = $1 AND org_id = $2
`;

export const SQL_SELECT_DOCUMENT_CHUNKS = `
  SELECT id, document_id, chunk_index, content, token_count, created_at
  FROM document_chunks
  WHERE document_id = $1
  ORDER BY chunk_index ASC
`;

export const SQL_DELETE_DOCUMENT = `
  DELETE FROM documents WHERE id = $1 AND org_id = $2 RETURNING id
`;

File: backend/app/sql/documents.py
from __future__ import annotations

from sqlalchemy import delete, select

from app.db.tables import document_chunks, documents

_DOC_COLS = [
    documents.c.id,
    documents.c.org_id,
    documents.c.connector_id,
    documents.c.external_id,
    documents.c.url,
    documents.c.title,
    documents.c.kind,
    documents.c.ext,
    documents.c.author_name,
    documents.c.author_email,
    documents.c.content_hash,
    documents.c.checksum,
    documents.c.content_type,
    documents.c.source_path,
    documents.c.source_last_modified_at,
    documents.c.source_permissions,
    documents.c.extraction_status,
    documents.c.extraction_error_code,
    documents.c.extraction_version,
    documents.c.chunking_version,
    documents.c.indexing_version,
    documents.c.word_count,
    documents.c.mtime,
    documents.c.indexed_at,
    documents.c.metadata,
]


def select_document_by_id(document_id, org_id):
    return select(*_DOC_COLS).where(
        documents.c.id == document_id, documents.c.org_id == org_id
    )


def select_document_chunks(document_id):
    return (
        select(
            document_chunks.c.id,
            document_chunks.c.document_id,
            document_chunks.c.chunk_index,
            document_chunks.c.content,
            document_chunks.c.token_count,
            document_chunks.c.created_at,
        )
        .where(document_chunks.c.document_id == document_id)
        .order_by(document_chunks.c.chunk_index.asc())
    )


def delete_document(document_id, org_id):
    return (
        delete(documents)
        .where(documents.c.id == document_id, documents.c.org_id == org_id)
        .returning(documents.c.id)
    )
